lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -31,8 +32,33 @@
 from ansible.module_utils.six import string_types
 from ansible.utils.version import SemanticVersion
 
+from collections.abc import Set
 from resolvelib import AbstractProvider
 
+PINNED_CANDIDATE_ATTRS = ('fqcn', 'ver', 'src', 'type')
+
+
+class PinnedCandidateRequests(Set):
+    def __init__(self, candidates):
+        self.candidates = set(candidates)
+
+    def __iter__(self):
+        return iter(self.candidates)
+
+    def __contains__(self, value):
+        for candidate in self.candidates:
+            for attr in PINNED_CANDIDATE_ATTRS:
+                if not hasattr(value, attr):
```

Comment:
Is this attempting to emulate
```suggestion
                if not isinstance(value, (Candidate, Requirement)):
```
?

Reply:
Also, looks like this class may belong to `ansible.galaxy.dependency_resolution.dataclasses` and use `_ComputedReqKindsMixin` in the check.

Reply:
Moreover, this check would be before both loops.

Reply:
`value` shouldn't be a Requirement (matches are Candidates). This conditional was just to be a little more resilient - but thinking longer about it, maybe I could just raise an exception instead since it would indicate a bug? Or it could just be removed. I updated it to be `if not isinstance(value, Candidate):` outside of the loops and and raise a ValueError if True - let me know what you think please.